### PR TITLE
Migrate to new padding rule

### PR DESCRIPTION
**Pull request by Al-Kindi-0**
## Describe your changes
Change the padding rule for RPO so that it uses the one in the xHash paper.

## Checklist before requesting a review
- Repo forked and branch created from `next` according to naming convention.
- Commit messages and codestyle follow [conventions](./CONTRIBUTING.md).
- Relevant issues are linked in the PR description.
- Tests added for new functionality.
- Documentation/comments updated according to changes.

### CODE REVIEW

File: stdlib/asm/crypto/hashes/native.masm
Comment (bobbinth) [inline stdlib/asm/crypto/hashes/native.masm:1]: I wonder if we should change the name of this module from `native` to `rpo`. On the one hand, it does mean that once we change to RPX, we might need to update the files which use `rpo` to `rpx` - but on the other hand, the change should be relatively simple and may be good to make sure we don't miss anything.

Also, looking at function names here, I wonder if we should:
1. Rename `state_to_digest` to something like `squeeze` or `squeeze_digest` or something similar (would this be more reflective what what we are doing here?)
2. Rename `hash_memory_even` to something like `absorb_double_words_from_memory`.

File: stdlib/asm/mem.masm
Comment (bobbinth) [inline stdlib/asm/mem.masm:89]: I'm wondering if we should change the semantics of this procedure to make it consistent with the `pipe_double_words_to_memory` procedure above. Specifically, this procedure could have the following interface:
```
#! Input:  [num_words, write_ptr, ...]
#! Output: [C, B, A, write_ptr', ...]
```
Then, we could squeeze the digest out in `pipe_preimage_to_memory` rather than here.

File: stdlib/asm/crypto/hashes/native.masm
Comment (bobbinth) [inline stdlib/asm/crypto/hashes/native.masm:92]: nit: instead of `push.0.0.0.0` we could use `padw` (won't change the cycle count, but is a bit cleaner to write).

Also, could we simplify this like so:
```
dropw dup.9 mem_loadw padw
```

File: stdlib/asm/mem.masm
Comment (bobbinth) [inline stdlib/asm/mem.masm:145]: minor nit: we could use `padw` instead of `push.0.0.0.0`.

File: assembly/src/assembler/instruction/crypto_ops.rs
Comment (Al-Kindi-0) [reply assembly/src/assembler/instruction/crypto_ops.rs:28]: Done

File: assembly/src/assembler/instruction/crypto_ops.rs
Comment (Al-Kindi-0) [reply assembly/src/assembler/instruction/crypto_ops.rs:37]: Not sure I follow, what are we duplicating in the first line?

File: stdlib/asm/crypto/hashes/native.masm
Comment (Al-Kindi-0) [reply stdlib/asm/crypto/hashes/native.masm:1]: 1. Changed to `squeeze_digest`.
2. Changed.

File: stdlib/asm/mem.masm
Comment (Al-Kindi-0) [reply stdlib/asm/mem.masm]: Done

File: stdlib/asm/crypto/hashes/native.masm
Comment (Al-Kindi-0) [reply stdlib/asm/crypto/hashes/native.masm]: Done

File: stdlib/asm/mem.masm
Comment (Al-Kindi-0) [reply stdlib/asm/mem.masm]: Done

File: stdlib/asm/mem.masm
Comment (bobbinth) [inline stdlib/asm/mem.masm:163]: Why did the cycle count increase by this much? We've added a call to `rpo::squeeze_digest` (which added 9 cycles) but `pipe_words_to_memory` should now be simpler, no? So, the delta should be about 2 cycles?

File: stdlib/tests/mem/mod.rs
Comment (bobbinth) [inline stdlib/tests/mem/mod.rs:131]: Should we replace this with `exec.rpo::squeeze_digest`. Also, maybe we should do a search&replace for `dropw swapw dropw` sequence to fix it in other places as well.

File: stdlib/tests/crypto/native.rs
Comment (bobbinth) [inline stdlib/tests/crypto/native.rs:1]: We should probably rename the file to `rpo.rs`.

**Review by bobbinth — APPROVED** [review]
Looks good! Thank you! I left a few more small comments inline.

Overall, let's hold off on merging this until we propagate the changes to `miden-base` (we can do that once we update `miden-base` to work with the current VM version).

File: stdlib/asm/mem.masm
Comment (Al-Kindi-0) [reply stdlib/asm/mem.masm:163]: The count is wrong. Here is a second attempt at it:
- `pipe_words_to_memory`: 41 + 9 * num_words / 2
- `squeeze_digest`: 9
- `movup.4 movdn.8`: 2
- `assert_eqw`: 10

The total is `62 + 9 * num_words / 2` for even number of words.
Similarly, the total is `79 + 9 * num_words / 2` for odd number of words.

File: stdlib/asm/crypto/hashes/rpo.masm
Comment (Al-Kindi-0) [reply stdlib/asm/crypto/hashes/rpo.masm:41]: done

File: stdlib/tests/mem/mod.rs
Comment (Al-Kindi-0) [reply stdlib/tests/mem/mod.rs]: done

File: stdlib/tests/crypto/native.rs
Comment (Al-Kindi-0) [reply stdlib/tests/crypto/native.rs:1]: changed

**Review by bobbinth — APPROVED** [review]
All looks good! Thank you!

File: stdlib/asm/crypto/hashes/rpo.masm
Comment (bobbinth) [inline stdlib/asm/crypto/hashes/rpo.masm:59]: One thing that I realized: I'm not sure the behavior of this procedure is entirely consistent with its naming/docs. Specifically, we assume that the number of elements to be hashed is a multiple of 4 - so, for example, it cannot be used to hash 9 elements.

Maybe a better approach would be to split it into 2 procedures:
```
#! Input:  [C, B, A, start_addr, num_words, ...]
#! Output: [C', B', A', end_addr, ...]
export.absorb_words_from_memory
```
The above procedure would be similar to `absorb_double_words_from_memory` but would handle the case of the odd number of words. (separately, I think we could modify the `absorb_double_words_from_memory` procedure to also take `num_words` instead of `end_addr`).

The second procedure would actually be for computing hash of memory:
```
#! Input:  [start_addr, num_elements, ...]
#! Output: [HASH, end_addr, ...]
export.hash_memory
```
One open question is whether we should assume that the last word is padded with zeros (in case the number of elements is not divisible by 4) or if we need to handle this in the procedure.

**Comment by bobbinth** [comment]
@Fumuran - could you merge the latest `next` into this branch?

**Comment by bobbinth** [comment]
@Fumuran could you update this branch to use `miden-crypto` v0.11.0 (I just published it to crates.io) and also merge `next` into this branch to make sure we don't have any merge conflicts?

File: core/Cargo.toml
Comment (bobbinth) [inline core/Cargo.toml:38]: Do we still need this line?

Comment (Fumuran) [reply core/Cargo.toml:38]: I wasn't sure, so decided to leave it. @Al-Kindi-0 Is the version of the crypto in your branch the same as v0.11?

Comment (Al-Kindi-0) [reply core/Cargo.toml:38]: I think after having merge the related PR in Miden crypto, we can now remove the above.

**Comment by bobbinth** [comment]
@Fumuran - this is basically done, right?

If so, before merging this, let's create a parallel PR in `miden-base` to migrate to the to this branch. I'm expecting that PR to be quite a bit of work because of the padding rule changes.

**Comment by Fumuran** [comment]
@bobbinth Should I prioritize migrating `miden-base` to the new padding rule over this issue (https://github.com/0xPolygonMiden/miden-base/issues/836)?

**Pull request merged**
